design/ErrorHandling.md

```diff
@@ -23,6 +23,147 @@ There are several types of WebGPU calls that get their errors handled differentl
+    this.adapter = null;
+    this.device = null;
+  }
+  async begin() {
+    try {
+      await this.initWebGPU();
+    } catch (e) {
+      console.error(e);
+      this.initFallback();
+    }
+  }
+  async initWebGPU() {
+    await this.ensureDevice();
+    // ... Upload resources, etc.
+  }
+  initFallback() { /* ... */ }
+  async ensureDevice() {
+    // Stop rendering. (If there was already a device, WebGPU calls made before
+    // the app notices the device is lost are okay - they are no-ops.)
+    this.device = null;
+
+    // Keep current adapter (but make a new one if there isn't a current one.)
+    // If we can't get an adapter, ensureDevice rejects and the app falls back.
+    await ensureAdapter();
+
+    try {
+      await ensureDeviceOnCurrentAdapter();
+      // Got a device.
+      return;
+    } catch (e) {
+      console.error("device request failed", e);
+      // That failed; try a new adapter entirely.
+      this.adapter = null;
+      // If we can't get a new adapter, it causes ensureDevice to reject and the app to fall back.
+      await ensureAdapter();
+      await ensureDeviceOnCurrentAdapter();
+    }
+  }
+  async ensureAdapter() {
+    if (!this.adapter) {
+      // If no adapter, get one.
+      // (If requestAdapter rejects, no matching adapter is available. Exit to fallback.)
+      this.adapter = await gpu.requestAdapter({ /* options */ });
+    }
+  }
+  async ensureDeviceOnCurrentAdapter() {
+    this.device = await this.adapter.requestDevice({ /* options */ });
+    this.device.lost.then((info) => {
+      // Device was lost.
+      console.error("device lost", info);
+      // Try to get a device again.
+      this.ensureDevice();
+    });
+  }
+}
+```
+
+### Case Studies
+
+*What signals should the app get, and when?*
+
+Two independent applications are running on the same webpage against two devices on the same adapter.
+The tab is in the background, and one device is using a lot of resources.
+ - The browser chooses to lose the heavier device.
+    - `device.lost` resolves, message = recovering device resources
+    - (App calls `createDevice` on any adapter, but it doesn't resolve yet.)
+ - Later, the browser might choose to lose the smaller device too.
+    - `device.lost` resolves, message = recovering device resources
+    - (App calls `createDevice` on any adapter, but it doesn't resolve yet.)
+ - Later, the tab is brought to the foreground. 
+    - Both `createDevice` Promises resolve.
+      (Unless the adapter was lost, in which case they would have rejected.)
+
+A page begins loading in a tab, but then the tab is backgrounded.
+ - On load, the page attempts creation of a device.
+    - `createDevice` Promise will resolve.
+
+A device's adapter is physically unplugged from the system (but an integrated GPU is still available).
+ - The same adapter, or a new adapter, is plugged back in.
+    - A later `requestAdapters` call may return the new adapter. TODO: `"gpuadapterschanged"`?
```

Comment:
Nit: If you haven't already done so, please open an issue to consider "gpuadapterschanged".  Upon resolution of the issue, the TODO would get removed or replaced with the new plan.